datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,211 @@
+import os
+from typing import Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkMetadata,
+    DocumentMetadataFilter,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding,
+)
+import json
+import psycopg
+from psycopg import sql
+
+from services.date import to_unix_timestamp
+
+# Read environment variables for Postgres
+
+POSTGRES_HOST = os.environ.get("POSTGRES_HOST", "localhost")
+POSTGRES_PORT = int(os.environ.get("POSTGRES_PORT", 5433))
+POSTGRES_USERNAME = os.environ.get("POSTGRES_USERNAME", "postgres")
+POSTGRES_PASSWORD = os.environ.get("POSTGRES_PASSWORD", "")
+POSTGRES_DATABASE = os.environ.get("POSTGRES_DATABASE", "pgml_development")
+POSTGRES_TABLENAME = os.environ.get("POSTGRES_TABLENAME", "chatgpt_datastore")
+POSTGRES_SYNCHRONOUS_COMMIT = os.environ.get(
+    "POSTGRES_SYNCHRONOUS_COMMIT", "off"
+).lower()
+
+assert POSTGRES_SYNCHRONOUS_COMMIT in ["on", "off"]
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+
+
+class PostgresDataStore(DataStore):
+    def __init__(self) -> None:
+        self.conn_params = {
+            "dbname": POSTGRES_DATABASE,
+            "user": POSTGRES_USERNAME,
+            "password": POSTGRES_PASSWORD,
+            "host": POSTGRES_HOST,
+            "port": POSTGRES_PORT,
+        }
+        conn = psycopg.connect(**self.conn_params)
+        # Insert the vector and text into the database
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
```

Comment:
`created_at`, `updated_at` should be structured `timestamp`s, although it's reasonable for metadata to override them on docs, not chunks. Is this another OpenAI "common" pattern?

Reply:
`created_at` is a `str` property of DocumentChunkMetadata. They have a helper function to convert `str` to `int` (Unix time stamp: seconds since epoch) 
```
DocumentChunkMetadata:
      title: DocumentChunkMetadata
      type: object
      properties:
        source:
          $ref: "#/components/schemas/Source"
        source_id:
          title: Source Id
          type: string
        url:
          title: Url
          type: string
        created_at:
          title: Created At
          type: string
        author:
          title: Author
          type: string
        document_id:
          title: Document Id
          type: string
          ```